
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Central Dogma: Decoding a Single Polypeptide &#8212; PyNotes in Agriscience</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Central Dogma: Decoding Multiple Polypeptides" href="central_dogma_multiple_polypeptides.html" />
    <link rel="prev" title="Complementary DNA sequence" href="matching_dna_strand.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  
  <h1 class="site-logo" id="site-title">PyNotes in Agriscience</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../landing-page.html">
   PyNotes in Agriscience
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/inspiration.html">
   Inspiration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/installing_software.html">
   Installating packages
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/reproducible_research.html">
   Reproducible Research
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/terminal_commands.html">
   Useful Terminal Commands
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/git_commands.html">
   Useful Git commands
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/my_first_repo.html">
   First Github repository
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/markdown_basics.html">
   Markdown basics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/latex_equations.html">
   LaTeX equations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/alt_python_libraries.html">
   Python libraries
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="hello_world.html">
   Hello world
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="help_resources.html">
   Python help
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Good coding practices
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/coding_style.html">
   Good Coding Habits
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../markdown/coding_guideline.html">
   Coding Guidelines
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Simple computations
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="calculator_examples.html">
   Calculator Examples
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="importing_modules.html">
   Importing modules
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="math_module.html">
   Math Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mass_volume_relationships.html">
   Soil Mass-Volume Relationships
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Random
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="random_module.html">
   Random module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="random_walk.html">
   Particle Random Walk
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="random_groups_of_people.html">
   Random group generator
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="random_plots.html">
   Random Plot Generator
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Data types structures
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="data_types.html">
   Data Types
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="data_structures.html">
   Data Structures
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="indexing_and_slicing_lists.html">
   Indexing and slicing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="strings.html">
   Strings
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="objects.html">
   Objects
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="classes.html">
   Classes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="inputs.html">
   Inputs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="input_wizard.html">
   Input Wizard
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="sedimentation_time.html">
   Sedimentation time
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Directory navigation
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="os_module.html">
   Operating system module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="glob_module.html">
   Glob Module
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Dates and Time
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="time_module.html">
   Time module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datetime_module.html">
   Datetime Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datetime_module.html#working-with-seconds">
   Working with seconds
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="leap_year.html">
   Leap years
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="day_of_year.html">
   Day of the Year
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="day_of_week.html">
   Day of the week
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Control flow
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="for_loops.html">
   For Loops
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="comparative_anatomy_for_loops.html">
   Comparative Anatomy of For Loops
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="while_loops.html">
   While loops
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="if_statement.html">
   If Statements
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="comparative_anatomy_if_statement.html">
   Comparative Anatomy of If Statements
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fibonacci_sequence.html">
   Fibonacci Sequence
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="sorting_numbers.html">
   Sorting Numbers
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="error_handling.html">
   Error Handling
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Functions
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="anonymous_functions.html">
   Lambda Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions.html">
   Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="comparative_anatomy_functions.html">
   Comparative Anatomy of Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="capillarity.html">
   Capillarity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pedotransfer_saxton_rawls.html">
   Pedotransfer Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_textural_class.html">
   Soil texture
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Strings
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="regular_expressions.html">
   Regular Expressions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="matching_dna_strand.html">
   Complementary DNA sequence
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Central Dogma: Decoding a Single Polypeptide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="central_dogma_multiple_polypeptides.html">
   Central Dogma: Decoding Multiple Polypeptides
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="morse_decoder.html">
   Morse code decoder
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="morse_encoder.html">
   Morse code encoder
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="morse_code_guessing_game.html">
   Morse code guessing game
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Plotting
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="plotting_matplotlib.html">
   Basic Plots with Matplotlib
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="plotting_seaborn.html">
   Seaborn
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="plotting_bokeh.html">
   Bokeh
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="plotting_holoviews.html">
   Holoviews
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interactive_apps.html">
   Interactive Application
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Numpy and Pandas
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="numpy_module.html">
   Numpy Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vapor_pressure_deficit.html">
   Vapor Pressure Deficit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="photoperiod_numpy_module.html">
   Photoperiod
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pandas_module.html">
   Pandas Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="import_tabular_data.html">
   Import Tabular Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="handy_functions.html">
   Handy functions for data science
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="precipitation_distribution.html">
   Precipitation Distribution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bird_counts.html">
   Exploring Bird Counts with Pandas and HoloViews
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Web data
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="url_ks_mesonet.html">
   Kansas Mesonet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="url_ok_mesonet.html">
   Oklahoma Mesonet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="url_uscrn.html">
   Climate Reference Network
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="url_nasa_api.html">
   Open NASA APIs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="coronavirus.html">
   Coronavirus COVID-19 Timeseries of Global Cases
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Statistics and curve fitting
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="statistics_module.html">
   Statistics module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="error_metrics.html">
   Error Metrics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="anova.html">
   Analysis of Variance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="group_least_variance.html">
   Group with the Least Variance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="stats_linear_regression.html">
   Linear Regression
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="sensor_calibration.html">
   Linear Regression: Sensor Calibration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="neutron_probe_calibration.html">
   Calibration of a Neutron Probe
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="anscombe_quartet.html">
   Anscombe’s Quartet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_respiration.html">
   Soil Respiration Rate
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="optimal_nitrogen_rate.html">
   Optimal Nitrogen Rate
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="maximum_return_to_nitrogen.html">
   Maximum Return to Nitrogen
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="proctor_test.html">
   Proctor test
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cdf_matching.html">
   CDF Matching Function
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="stats_multiple_linear_regression.html">
   Multiple Linear Regression Analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="frontier_function.html">
   Frontier Production Function
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="stats_curve_fitting_exp.html">
   Curve Fitting: Exponential Decay
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cosmos_calibration.html">
   Neutron Counts to Volumetric Water Content
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="stats_curve_fitting_wave.html">
   Curve Fitting: Sinusoid
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_water_retention_curve.html">
   Soil Water Retention Curve
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="atmospheric_carbon_dioxide.html">
   Atmospheric Carbon Dioxide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="birthday_paradox.html">
   Birthday Paradox
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="neural_network_regression.html">
   Neural Network Regression
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="neural_network_classification.html">
   Neural Network Classification
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Problems in 1D
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="fill_missing_data.html">
   Fill Missing Soil Moisture Timeseries
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="denoise_passive_neutron_counts.html">
   Denoising Cosmic-Ray Neutron Signal
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="wheat_potential_biomass.html">
   Modeling Wheat Biomass
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="wheat_phenology.html">
   Wheat phenology model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="markov_chain.html">
   Markov Chains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="antecedent_precipitation_index.html">
   Antecedent Precipitation Index
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="markov_chain_soil_moisture.html">
   Modeling Soil Moisture
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rainfall_generator.html">
   Rainfall Generator
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="precipitation_metrics.html">
   Precipitation Metrics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="atmospheric_water_deficit.html">
   Atmospheric Water Deficit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="drydowns_from_timeseries.html">
   Soil Moisture Drydowns
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="air_temperature.html">
   Air Temperature
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="solar_radiation.html">
   Solar Radiation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_temperature.html">
   Understanding Soil Temperature
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_temperature_model.html">
   Soil Temperature Model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_water_storage.html">
   Soil Water Storage
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="exponential_filter.html">
   Rootzone Soil Water from Surface Observations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="plant_available_water.html">
   Plant Available Water
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mixing_model.html">
   Mixing Solute Problem
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="newton_law_cooling.html">
   Newton’s law of cooling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="lotka_volterra.html">
   Lotka-Volterra: Predator–Prey Model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="leaky_bucket.html">
   Leaky buckets
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="thermal_time.html">
   Thermal Units
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_temperature_model.html">
   Soil Temperature Model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="root_depth.html">
   Root growth
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_water_balance.html">
   Soil Water Balance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="evaporation.html">
   Soil Evaporation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="evapotranspiration.html">
   Potential Evapotranspiration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="curve_number.html">
   Runoff
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Image analysis
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="canopeo.html">
   Green Canopy Cover
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="count_seeds.html">
   Counting seeds
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="segmented_lake.html">
   Patch analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_cracks_detection.html">
   Soil Cracks
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Mapping
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="nearest_station.html">
   Find Nearest Monitoring Station
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="inverse_distance_weighting.html">
   Inverse Distance Weighting Interpolation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interpolation.html">
   Spatial Interpolation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="raster_nasa_worldview.html">
   Rasters from NASA Worldview Snapshots
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="land_cover.html">
   Land Cover Classification
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="denoise_yield_monitor.html">
   Denoise Yield Monitor Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="raster_change_crs.html">
   Change Coordinate Reference System
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="field_random_samples.html">
   Field Random Samples
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="euclidean_distance.html">
   Efficient Euclidean Distance
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Numerical problems 2D
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="game_of_life.html">
   Conway’s Game of Life
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="wildfire_spread.html">
   Wildfire Model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="reaction_diffusion.html">
   Gray-Scott Model of Reaction-Diffusion
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="disease_spread.html">
   Disease Spread
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Clustering
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="clustering_environments.html">
   Environmental Agroregions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="soil_water_deficit_classification.html">
   Clustering Time Series
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="field_management_zones.html">
   Field Management Zones
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Symbolic math
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="sympy_module.html">
   Sympy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="logistic_growth.html">
   Logistic Growth
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="population_growth.html">
   Population Growth
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/notebooks/central_dogma_single_polypeptide.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/soilwater/pynotes-agriscience/master?urlpath=tree/notebooks/central_dogma_single_polypeptide.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#assumptions">
   Assumptions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#transcription">
   Transcription
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#translation">
   Translation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dna-in-action">
   DNA in action
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   References
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="central-dogma-decoding-a-single-polypeptide">
<h1>Central Dogma: Decoding a Single Polypeptide<a class="headerlink" href="#central-dogma-decoding-a-single-polypeptide" title="Permalink to this headline">¶</a></h1>
<p><em>The DNA is the carrier of genetic information and its simplest form can be viewed as a long string of characters that encode biological features and attributes</em></p>
<p>The central dogma of molecular biology consists of a two step process to convert the information contained in DNA sequences into proteins. These steps are transcription (DNA into messenger RNA) and translation (messenger RNA into aminoacids). The RNA polymerase is an enzyme that participates in the transcription process, while ribosomes are large molecular structures that translate messenger RNA into a chain of aminoacids to form polypetides and proteines.</p>
<img alt="https://www.nature.com/scitable/content/ne0000/ne0000/ne0000/ne0000/105292327/44350_36a.jpg" src="https://www.nature.com/scitable/content/ne0000/ne0000/ne0000/ne0000/105292327/44350_36a.jpg" />
<p><strong>Figure: A gene is expressed through the processes of transcription and translation.
During transcription, the enzyme RNA polymerase (green) uses DNA as a template to produce a pre-mRNA transcript (pink). The pre-mRNA is processed to form a mature mRNA molecule that can be translated to build the protein molecule (polypeptide) encoded by the original gene.
Source: Nature Education.</strong></p>
<p>You can learn more by reading an excellent educational article by <a class="reference external" href="https://www.nature.com/scitable/topicpage/translation-dna-to-mrna-to-protein-393/">Suzanne Clancy and William Brown (2008)</a></p>
<p>The <strong>goal</strong> of this exercise is to use Python to emulate both the RNA polymerase and a ribosome to decode the aminoacids present in a sequence of DNA. To accomplish this goal we will need a DNA sequence and a lookup table to convert codons into aminoacids. I transcribed the information in the table below into a JSON file.</p>
<img alt="https://www.nature.com/scitable/content/ne0000/ne0000/ne0000/ne0000/105292568/44376_38b.jpg" src="https://www.nature.com/scitable/content/ne0000/ne0000/ne0000/ne0000/105292568/44376_38b.jpg" />
<p><strong>Figure: The large ribosomal subunit binds to the small ribosomal subunit to complete the initiation complex.
The initiator tRNA molecule, carrying the methionine amino acid that will serve as the first amino acid of the polypeptide chain, is bound to the P site on the ribosome. The A site is aligned with the next codon, which will be bound by the anticodon of the next incoming tRNA. Source: Nature Education.</strong></p>
<p>Codons refer to the triplets in the mRNA. The triplets in the sequence of transfer RNA (tRNA) are called anti-codons and are complementary to the codons in the mRNA. In the figure above, <code class="docutils literal notranslate"><span class="pre">AUG</span></code> (in the 5’ to 3’ mRNA) is the codon for Methionine, while <code class="docutils literal notranslate"><span class="pre">UAC</span></code> (in the tRNA) is the anti-codon for Methionine</p>
<p>There are total of 23 essential aminoacids that encoded in codons consisting of three bases. Some codons are dedicated at instructing the ribosomes the START or STOP of the sequence. The genetic code includes 64 possible permutations, or combinations, of three-letter nucleotide sequences that can be made from the four nucleotides. Of the 64 codons, 61 represent amino acids, and three are stop signals. For example, the codon CAG represents the amino acid glutamine, and TAA is a stop codon. The genetic code is described as degenerate, or redundant, because a single amino acid may be coded for by more than one codon. When codons are read from the nucleotide sequence, they are read in succession and do not overlap with one another.</p>
<div class="section" id="assumptions">
<h2>Assumptions<a class="headerlink" href="#assumptions" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>For simplicity we will ignore additional portions of DNA such as promoter and a terminator regions that play an important role during transcription, but that do not contribute the sequence of the polypeptide.</p></li>
<li><p>All bases will be represented in upper case for consistency.</p></li>
<li><p>We will assume that the ribosome is made of a single unit and contains a single activation site. So basically a simplified version of the figure above.</p></li>
</ul>
<p>As usual we will start by importing the necessary modules and by loading all the data for the exercise.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import modules</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read codon-aminoacids lookup table</span>
<span class="c1"># Note that the table mathes aminoacids to codons (not the anti-codon)</span>
<span class="n">lookup</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../datasets/codon_aminoacids.csv&#39;</span><span class="p">)</span>
<span class="n">lookup</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>codon</th>
      <th>letter</th>
      <th>aminoacid</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AAA</td>
      <td>K</td>
      <td>Lysine</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AAC</td>
      <td>N</td>
      <td>Asparagine</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAG</td>
      <td>K</td>
      <td>Lysine</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAU</td>
      <td>N</td>
      <td>Asparagine</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ACA</td>
      <td>T</td>
      <td>Threonine</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Random DNA sequence</span>
<span class="c1"># Source: https://www.bioinformatics.org</span>
<span class="n">DNA</span> <span class="o">=</span> <span class="s1">&#39;tacctatttcactgccgtccgttgcactacgaacggaagccgtgctcagaccaacacgtccagcaacaaagaact&#39;</span>
<span class="n">DNA</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">DNA</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TACCTATTTCACTGCCGTCCGTTGCACTACGAACGGAAGCCGTGCTCAGACCAACACGTCCAGCAACAAAGAACT
</pre></div>
</div>
</div>
</div>
<p>I decided to keep it simple, but you can create longer and more diverse DNA sequences using this <a class="reference external" href="https://www.bioinformatics.org/sms2/random_coding_dna.html">random coding DNA tool</a>. For clarity, notice that this tool already provides the complementary DNA strand. Since in this exercise we are also implementing the transcription step I reversed the DNA sequence obtained from the website. Here is the original DNA strand that I obtained from the online tool: <code class="docutils literal notranslate"><span class="pre">atggataaagtgacggcaggcaacgtgatgcttgccttcggcacgagtctggttgtgcaggtcgttgtttcttga</span></code>, which I reversed using the code from the <code class="docutils literal notranslate"><span class="pre">complementaty</span> <span class="pre">DNA</span> <span class="pre">sequence</span></code> exercise. If you don’t do this step, the code will still decode aminoacids, but the resulting polypeptide will not look right (i.e. Methionine will not be the first aminoacid). I simply reversed the DNA strand to closely match textbook examples and available material online.</p>
</div>
<div class="section" id="transcription">
<h2>Transcription<a class="headerlink" href="#transcription" title="Permalink to this headline">¶</a></h2>
<p>Given a sequence of DNA bases we need to find the complementary strand. The catch here is that we also need to account for the fact that the base <code class="docutils literal notranslate"><span class="pre">thymine</span></code> is replaced by the base <code class="docutils literal notranslate"><span class="pre">uracil</span></code> in RNA.</p>
<p>To check for potential typos in the sequence of DNA or to prevent that the user feeds a sequence of mRNA instead of DNA to the transcription function, we will use the <code class="docutils literal notranslate"><span class="pre">raise</span></code> statement, which will automatically stop and exit the <code class="docutils literal notranslate"><span class="pre">for</span></code> loop and throw a custom error message if the code finds a base a base other than A,T,C, or G. The location of the <code class="docutils literal notranslate"><span class="pre">raise</span></code> statement is crucial since we only want to trigger this action if a certain condition is met (i.e. we find an unknown base). So, we will place the <code class="docutils literal notranslate"><span class="pre">raise</span></code> statement inside the <code class="docutils literal notranslate"><span class="pre">if</span></code> statement within the <code class="docutils literal notranslate"><span class="pre">for</span></code> loop. We will also return the location in the sequence of the unknown base using the <code class="docutils literal notranslate"><span class="pre">find()</span></code> method.</p>
<p>The error catching method described above is simple and practical for small applications, but it has some limitations. For instance, we cannot identify whether there are more than one unknwon bases and we cannot let the user know the location of all these bases. Nonetheless, this is a good starting point.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">transcription</span><span class="p">(</span><span class="n">DNA</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Docstring: Function that converts a strand of DNA into </span>
<span class="sd">    a strand of messenger RNA based on the following nucleotide paris</span>
<span class="sd">    A-U, T-A, C-G, G-C.</span>
<span class="sd">    </span>
<span class="sd">    Input must be a single strand of DNA in string format.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">mRNA</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span> <span class="c1"># initialize strand of messenger RNA</span>
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">DNA</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">:</span>
            <span class="n">mRNA</span> <span class="o">+=</span> <span class="s1">&#39;U&#39;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span>
            <span class="n">mRNA</span> <span class="o">+=</span> <span class="s1">&#39;A&#39;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span>
            <span class="n">mRNA</span> <span class="o">+=</span> <span class="s1">&#39;G&#39;</span>
        <span class="k">elif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span>
            <span class="n">mRNA</span> <span class="o">+=</span> <span class="s1">&#39;C&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">errorMessage</span> <span class="o">=</span> <span class="s1">&#39;Unknown base: &#39;</span> <span class="o">+</span> <span class="n">base</span> <span class="o">+</span> <span class="s1">&#39; found at: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">DNA</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">base</span><span class="p">))</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="n">errorMessage</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">mRNA</span>
</pre></div>
</div>
</div>
</div>
<p>After solving the problem with a short script, we could easily convert the script into a function. This will allow us to use the function multiple times without re-writing the instructions.</p>
</div>
<div class="section" id="translation">
<h2>Translation<a class="headerlink" href="#translation" title="Permalink to this headline">¶</a></h2>
<p>We are at a breakpoint. We loaded the lookup table and we have the mRNA sequence. Our next step is to solve the translation problem. Basically we need to find a way to search the matching aminoacid for any given codon. If we we find a way to do this, then we can implement it into a <code class="docutils literal notranslate"><span class="pre">for</span></code> loop and apply to the entire mRNA sequence. We will start simple, I want to show you the value of a trivial example.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Test that we can match a codon an retrieve the aminoacid</span>
<span class="n">aa_idx</span> <span class="o">=</span> <span class="n">lookup</span><span class="o">.</span><span class="n">codon</span> <span class="o">==</span> <span class="s1">&#39;AAA&#39;</span> <span class="c1"># AAA is a possible codon from the lookup the table</span>
<span class="n">lookup</span><span class="o">.</span><span class="n">aminoacid</span><span class="p">[</span><span class="n">aa_idx</span><span class="p">]</span>       <span class="c1"># Use the resulting boolean vector to obtain the aminoacid</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0    Lysine
Name: aminoacid, dtype: object
</pre></div>
</div>
</div>
</div>
<p>We cracked the translation problem! The statement can successfully match and retrieve the name of the matching aminoacid given a codon of three bases. This is exactly what we mean with breaking down problems. It can be extremely rewarding to solve small bits of code and will prevent you from writing long chunks of code that are hard to troubleshoot. With practice you will be able to learn how to breakdown large problems into small pieces that can be solved as a logical sequence of smaller problems.</p>
<p>Let’s not get too excited, while the answer is correct, the format does not seem entirely correct. We did not obtain a string. The answer also contains additional information (e.g. Name and dtype) about the object. We need to keep in mind that we need to be able to use the resulting string to do something with it (e.g. create a chain of aminoacid names, which will resemble a polypeptide).</p>
<p>So here is a better solution to extract the information we really need:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lookup</span><span class="o">.</span><span class="n">aminoacid</span><span class="p">[</span><span class="n">aa_idx</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># This syntax access the string within the Pandas object</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;Lysine&#39;
</pre></div>
</div>
</div>
</div>
<p>Now we can proceed to translate the mRNA into a chain of aminoacids</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">translation</span><span class="p">(</span><span class="n">mRNA</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Docstring: Function that trsnlates a strand of mRNA into </span>
<span class="sd">    a sequence of aminoacids, which is know as a polypeptide.</span>
<span class="sd">    </span>
<span class="sd">    Input must be a single strand of mRNA in string format.</span>
<span class="sd">    &#39;&#39;&#39;</span>
        
    <span class="c1"># Initial conditions of our ribosome</span>
    <span class="n">polypeptide</span> <span class="o">=</span> <span class="p">[];</span>

    <span class="c1"># Iterate over each codon triplet. We will use a step of 3 in the loop.</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">mRNA</span><span class="p">)</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">):</span>
        <span class="n">codon</span> <span class="o">=</span> <span class="n">mRNA</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="p">]</span> <span class="c1"># Add 3 to avoid overlapping the bases between iterations.</span>
        <span class="n">aminoacid_idx</span> <span class="o">=</span> <span class="n">lookup</span><span class="o">.</span><span class="n">codon</span> <span class="o">==</span> <span class="n">codon</span> <span class="c1"># Match current codon with all codons in lookup table</span>
        <span class="n">aminoacid</span> <span class="o">=</span> <span class="n">lookup</span><span class="o">.</span><span class="n">aminoacid</span><span class="p">[</span><span class="n">aminoacid_idx</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">polypeptide</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aminoacid</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">polypeptide</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="dna-in-action">
<h2>DNA in action<a class="headerlink" href="#dna-in-action" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mRNA</span> <span class="o">=</span> <span class="n">transcription</span><span class="p">(</span><span class="n">DNA</span><span class="p">)</span>
<span class="n">polypeptide</span> <span class="o">=</span> <span class="n">translation</span><span class="p">(</span><span class="n">mRNA</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;DNA sequence:&#39;</span><span class="p">,</span><span class="n">DNA</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;mRNA sequence:&#39;</span><span class="p">,</span><span class="n">mRNA</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Polypeptide:&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">polypeptide</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DNA sequence: TACCTATTTCACTGCCGTCCGTTGCACTACGAACGGAAGCCGTGCTCAGACCAACACGTCCAGCAACAAAGAACT
mRNA sequence: AUGGAUAAAGUGACGGCAGGCAACGUGAUGCUUGCCUUCGGCACGAGUCUGGUUGUGCAGGUCGUUGUUUCUUGA
Polypeptide: Methionine-Aspartic_acid-Lysine-Valine-Threonine-Alanine-Glycine-Asparagine-Valine-Methionine-Leucine-Alanine-Phenylalanine-Glycine-Threonine-Serine-Leucine-Valine-Valine-Glutamine-Valine-Valine-Valine-Serine-Stop
</pre></div>
</div>
</div>
</div>
<p>Despite some simplifications, our exercise makes sense. Methionine is the first aminoacid of new proteins (although it can be removed in subsequent steps) and our last codon resulted to be a stop codon. The resulting sequence is our polypeptide!</p>
<p>If you deal with DNA and mRNA seqeuences on a daily basis, consider creating a module with transcription and translation functions that you can readily use in multiple projects.</p>
<p>Our code can only handle a single sequence encoding a single polypeptide. Below I leave a DNA sequence encoding 3 polypeptides. Can you write a function capable of decoding a DNA sequence of multiple polypeptides, each with different number of aminoacids?</p>
<p><code class="docutils literal notranslate"><span class="pre">TACTCGTCACAGGTTACCCCAAACATTTACTGCGACGTATAAACTTACTGCACAAATGTGACT</span></code></p>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p>Clancy, S. and Brown, W., 2008. Translation: DNA to mRNA to protein. Nature Education, 1(1), p.101.</p>
<p>Stothard P. 2000. The Sequence Manipulation Suite: JavaScript programs for analyzing and formatting protein and DNA sequences. Biotechniques 28:1102-1104.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./notebooks"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="matching_dna_strand.html" title="previous page">Complementary DNA sequence</a>
    <a class='right-next' id="next-link" href="central_dogma_multiple_polypeptides.html" title="next page">Central Dogma: Decoding Multiple Polypeptides</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Andres Patrignani<br/>
        
            &copy; Copyright 2021.<br/>
          <div class="extra_footer">
            <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/">Licensed under CC BY-NC-SA</a>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>